draft-gpew-priv-ppm.md

```diff
@@ -947,26 +947,50 @@ aggregator has validated and which fall in the batch interval of the request.
-R1.nonce`, then we say that `R2` does not follow `R1`.
-
-To prevent replay attacks, each aggregator ensures that each report it
-aggregates follows the previously aggregated report. To prevent the adversary
-from tampering with the ordering of reports, honest clients incorporate the
-ordering-sensitive parameters `(time, nonce)` into the AAD for HPKE encryption.
-Note that this strategy may result in dropping reports that happen to have the
-same timestamp and nonce value.
+about the client's measurement, violating the privacy goal of PPM. To
+prevent such replay attacks, this specification requires the aggregators to
+detect and filter out replayed reports.
+
+To detect replay attacks, each aggregator keeps track of the set of nonces
+pertaining to reports that were previously processed for a given task. If the
+leader receives a report from a client whose nonce is in this set, it simply
+ignores it. Similarly, a helper ignores an encrypted input share whose nonce is
+in this set.
+
+A malicious aggregator may attempt to force a replay by replacing the nonce
+generated by the client with a nonce its peer has not yet seen. To prevent this,
+clients incorporate the nonce into the AAD for HPKE encryption, ensuring that
+the input share is only recovered if the aggregator is given the correct nonce.
+(See {{upload-request}}.)
 
 Aggregators prevent the same report from being used in multiple batches (except
 as required by the protocol) by only responding to valid collect requests, as
 described in {{batch-parameter-validation}}.
 
+The size of the nonce set is unbounded, and depending the task, can grow quite
+large. To ensure they do not need to store these indefinitely, each aggregator
+MAY at its own discretion ignore reports with timestamps that are too far in the
+past, too far in the future, or would fall in a batch interval that has already
+been aggregated (see {{batch-parameter-validation}}).
+
+[OPEN ISSUE: Allowing aggregators to ignore reports at-will is likely to lead
+to reports getting dropped unnecessarily. Ideally there would be a way to force
+the aggregators to make the same decision about the time intervals for which
+nonces are tracked and for which nonces are ignored. There are two ways to this:
+
+1. Pick a duration `d_min` for reports too far in the future and a duration
+   `d_max` for reports too far in the past, and ignore a report if `now() - time
+   > d_dmin` or `time - now() > d_max`, where `time` is the report timestamp.
+   This requires the aggregators to have a roughly synchronized clock.
+
+2. @martinthomson made the following suggestion, which doesn't require clock
+   sync. The leader and helper could periodically negotiate the time `t_min` in
+   the past before which reports are deemed too old and the time `t_max` in the
+   future after which reports are deemed too new. This is fairly straight
+   forward to piggy back on aggregate requests, but would get a bit trickier in
+   the case of multiple helper.]
```

Comment:
I don't see how this is difficult.  The leader starts a run and tells helpers what t_min and t_max are as part of that.

Reply:
Shouldn't be too bad, but you do have to negotiate `t_min, t_max` that works for all of the helpers. You have to make sure that whatever `t_min` you pick is newer than than the oldest value that is tolerated by any aggregator. (Similarly for `t_max`, you'd have to pick it so that it's older than the newest value tolerated by any aggregator.)

Reply:
This just seems unspecified.